specification/common-types/resource-management/v4/customermanagedkeys.json

```diff
@@ -0,0 +1,70 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "version": "4.0",
+    "title": "Common types"
+  },
+  "paths": {},
+  "definitions": {
+    "encryption": {
+      "type": "object",
+      "description": "(Optional) Discouraged to include in resource definition. Only needed where it is possible to disable platform (AKA infrastructure) encryption. Azure SQL TDE is an example of this. Values are enabled and disabled.",
+      "properties": {
+        "infrastructureEncryption": {
+          "type": "string",
+          "enum": [
+            "enabled",
+            "disabled"
+          ],
+          "x-ms-enum": {
+            "name": "InfrastructureEncryption",
+            "modelAsString": true
+          },
+          "description": "Values are enabled and disabled."
+        },
+        "customerManagedKeyEncryption": {
+          "description": "All Customer-managed key encryption properties for the resource.",
+          "$ref": "#/definitions/customerManagedKeyEncryption"
+        }
+      }
+    },
+    "customerManagedKeyEncryption": {
+      "type": "object",
+      "description": "All Customer-managed key encryption properties for the resource.",
+      "properties": {
+        "keyEncryptionKeyIdentity": {
+          "type": "object",
+          "description": "All identity configuration for Customer-managed key settings defining which identity should be used to auth to Key Vault.",
+          "properties": {
+            "identityType": {
+              "type": "string",
+              "enum": [
+                "systemAssignedIdentity",
+                "userAssignedIdentity",
+                "none"
+              ],
+              "x-ms-enum": {
+                "name": "IdentityType",
+                "modelAsString": true
+              },
+              "description": "Values can be systemAssignedIdentity or userAssignedIdentity"
+            },
+            "userAssignedIdentityResourceId": {
+              "type": "string",
+              "description": "user assigned identity to use for accessing key encryption key Url. Ex: /subscriptions/fa5fc227-a624-475e-b696-cdd604c735bc/resourceGroups/<resource group>/providers/Microsoft.ManagedIdentity/userAssignedIdentities/myId. Mutually exclusive with identityType systemAssignedIdentity and none."
+            },
+            "delegatedIdentityClientId": {
```

Comment:
@jackrichins please note we also added this field (although it isn't specified in https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-contracts.md#customer-managed-key-encryption).

Reply:
Why is this needed? Why not use a User Assigned Identity?

Reply:
We are using HOBO-V2 (as defined [here](https://armwiki.azurewebsites.net/authorization/managed_identities/hosted_on_behalf_of_v2/Overview.html) and [here](https://armwiki.azurewebsites.net/authorization/managed_identities/hosted_on_behalf_of_v2/BackingResourceProviderOnboarding.html) to accept delegated identities.
We need to use chained delegation with CMK - our RP needs to accept a delegated identity from another RP and use this identity for CMK.

Reply:
I asked Storage how they handle delegatedIdentity and they responded with the following:
We have an AFEC – [Microsoft.Storage](http://microsoft.storage/)\AllowSharedIdentities that first party services should onboard onto in order to use the property. Only subscriptions in the Microsoft/AME tenant are whitelisted to register the AFEC. Also, the property is not exposed to any external customers and is not part of the external REST API specifications.
To learn more, Manu Suresh Rao [manusrao@microsoft.com](mailto:manusrao@microsoft.com) is the engineer who owns it for Storage.

Reply:
important to mention that unlike storage, we are RPaaS-based RP. which means RPaaS validates the request body for us based on our swagger. unlike storage, we can't bypass properties without declaring them.
UserRP can not opt out from schema validation by RPaaS (and we also don't want to of course 😄).

btw, AFAIU storage no longer requires AFEC for encryption scopes with delegation.

Reply:
as a side note, i think it's good thing to document the schema, and to do it via common types. it enforces consistent APIs and ease internal teams to work with other RPs.
ideally will be great to have internal common types, but in the absence of such internal place, common type is the best option IMHO.